### ISSUE TITLE
Where to host cs.k8s.io

### ISSUE THREAD

**Issue by ameukam** [kubernetes/k8s.io#2182]
https://cs.k8s.io is running on a baremetal server [provided](https://github.com/cncf/cluster/issues/67)  by Equinix Metal(ex Packet) under CNCF budget and operated until now by @dims. 

The question was asked about whether or not we should host CodeSearch on [aaa](https://github.com/kubernetes/k8s.io/blob/main/running-in-community-clusters.md#where-are-the-clusters) cluster. 

Ref: https://kubernetes.slack.com/archives/CCK68P2Q2/p1615204807111900?thread_ts=1615189697.108500&cid=CCK68P2Q2

Issue open to track the discussions and the consensus about this.

**Comment by jimdaga** [comment]
I took a stab at onboarding codesearch; @spiffxp could I get your input? I want to make sure I didn't miss anything. 
I want to stage all the infra, and get it deployed via prow first. Then we can follow up with another PR to cut-over DNS when we are ready.

https://github.com/kubernetes/k8s.io/pull/2513
https://github.com/kubernetes/test-infra/pull/23201

I could also work on adding the docker build logic after, but I haven't worked in that repo yet so I'll have to do some digging. 

cc @dims

**Comment by justaugustus** [comment]
What about using https://sourcegraph.com/kubernetes to minimize the maintenance burden here?
This is something I suggested to @dims in the past, but didn't have the bandwidth to do at the time.

**Comment by dims** [comment]
choices are:
1. leave things where they are 
2. move to k8s wg infra
3. redirect to cs.k8s.io to sourcegraph

- i have been taking care of 1 already for a while with minimal downtime, so i am ok with continuing to do so
- if someone wants to do 2, i am happy to work with help, show how things are setup and we can shut down the equinix vm
- i personally don't like option 3, i love the the hound UX, if the consensus is we should go with 3, that is fine with me. I am happy to run a personal instance on a custom domain for myself (community is welcome to use)

if i missed any other options, please feel free to chime in.

**Comment by jimdaga** [comment]
FYI: If choice #2 is picked, my two PRs are pretty much ready to stage codesearch in the `aaa` cluster. There are a few small things that need to happen after the merge, but that's documented in my PRs. 
- https://github.com/kubernetes/k8s.io/pull/2513
- https://github.com/kubernetes/test-infra/pull/23201

**Comment by dims** [comment]
thanks @jimdaga 

+1 to give #2 a shot. will let Aaron and Arnaud to review and merge all 3 PRs

**Comment by nikhita** [comment]
@ameukam what is remaining here?

**Comment by ameukam** [comment]
> @ameukam what is remaining here?

Deploy a canary instance from https://github.com/kubernetes/k8s.io/pull/2513. Once we have confidence with that instance we can rollout a prod instance.

**Comment by Priyankasaggu11929** [comment]
/assign

@nikhita, I'm interested in helping with setting up a canary instance.

**Comment by Priyankasaggu11929** [comment]
Post-merge checklist item from PR https://github.com/kubernetes/k8s.io/pull/2513 that need working on:
- [ ] Publish `cs-fetch-repo` docker image (Open PR: https://github.com/kubernetes/test-infra/pull/25576)
- [ ] Update deployment to use deployed docker image (using a temp image for now)
- [X] After testing is completed, cutover DNS to new K8s hosted IP. (Done by https://github.com/kubernetes/k8s.io/pull/3416)

@pmgk07, once we're done with having https://github.com/kubernetes/test-infra/pull/25576 merged for adding `cs-fetch-repos` image under k8s infra, the next step would be updating the [codesearch/deployment.yaml#L27](https://github.com/kubernetes/k8s.io/blob/main/apps/codesearch/deployment.yaml#L27) to use above hosted image.

**Comment by ameukam** [comment]
> Update deployment to use deployed docker image (using a temp image for now)

@Priyankasaggu11929 Let's give @jimdaga the final call about this. There are possible changes that need to be added the Docker image.

**Comment by jimdaga** [comment]
Now that https://github.com/kubernetes/k8s.io/pull/3492 is merged, I see codesearch is deployed in the cluster! 

However, it looks like the init containers are crashing:
```
kubectl get pods -n codesearch
NAME                         READY   STATUS                  RESTARTS   AGE
codesearch-5b975d449-lgm9b   0/1     Init:CrashLoopBackOff   8          19m
codesearch-5b975d449-zzqkl   0/1     Init:CrashLoopBackOff   8          19m
```
I'm out of the office right now, so I can't do a full debug. But it does seem like something needs fixing :( 
(I also don't have access to view pod logs, so not sure how to get that)

**Comment by Priyankasaggu11929** [comment]
> Let's give @jimdaga the final call about this. There are possible changes that need to be added the Docker image.

+1. Yes 🙂 

There's also an error for decoding ingress in the [build-logs](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/post-k8sio-deploy-app-codesearch/1502042402167197696) of the `post-k8sio-deploy-app-codesearch` job. 

I've raised a minor patch fix: https://github.com/kubernetes/k8s.io/pull/3502

**Comment by ameukam** [comment]
> Now that #3492 is merged, I see codesearch is deployed in the cluster!
> 
> However, it looks like the init containers are crashing:
> 
> ```
> kubectl get pods -n codesearch
> NAME                         READY   STATUS                  RESTARTS   AGE
> codesearch-5b975d449-lgm9b   0/1     Init:CrashLoopBackOff   8          19m
> codesearch-5b975d449-zzqkl   0/1     Init:CrashLoopBackOff   8          19m
> ```
> 
> I'm out of the office right now, so I can't do a full debug. But it does seem like something needs fixing :( (I also don't have access to view pod logs, so not sure how to get that)

You can use GCP Logging console for the logs: https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.namespace_name%3D%22codesearch%22;cursorTimestamp=2022-03-11T06:20:53.646489047Z?project=kubernetes-public.

I did a quick research based on the logs and it suggested the issue may be related to the architecture of the Docker image.

```shell
 skopeo inspect docker://jdagostino2/codesearch-fetch:0.1.7 | jq .Architecture
"arm64"
``` 
The image seems to be built using a `arm64` processor but the GKE nodes are `amd64`. We should try to switch to `gcr.io/k8s-staging-infra-tools` and see what's happening.

I took a quick look.

**Comment by pmgk07** [comment]
@ameukam I unknowingly added fixes #xyz in my PR which led the k8s-ci-robot to close this issue. Feel free to reopen this issue if there's anything pending.

**Comment by ameukam** [comment]
https://cs-canary.k8s.io is now up and running. We should spread the word about his existence and think about we will flip `cs.k8s.io` to `aaa` GKE cluster

**Comment by dims** [comment]
sounds like a great plan @ameukam ! we can flip ASAP. i will leave the other one running for a week or two just in case we have a problem

**Comment by jimdaga** [comment]
The one last change we need before we flip is a job that restarts the deployment nightly to pick up any changes.
The way I set up the deployment there should be no downtime while the new pods are coming up.

**Comment by jimdaga** [comment]
Hopefully have the final two PRs needed to consider this "go-live" ready. 

Looking for an LGTM for these two:

- https://github.com/kubernetes/k8s.io/pull/3679 
- https://github.com/kubernetes/test-infra/pull/26105

**Comment by nikhita** [comment]
/remove-lifecycle stale

@jimdaga do you have cycles to address the review on https://github.com/kubernetes/k8s.io/pull/3679?

**Comment by upodroid** [comment]
Can we explore deprecating this in favour of GitHub Code Search?

https://cs.github.com/
https://github.com/features/code-search

**Comment by ameukam** [comment]
Not really. one issue is that Github CS requires auth to use it while we have anonymous queries using `cs.k8s.io` (e.g. https://go.k8s.io/owners/dims).

**Comment by BenTheElder** [comment]
/lifecycle frozen
closing this is not helpful, unless we've also shut down the existing infra.

**Comment by SohamChakraborty** [comment]
I think this is now ready for migration from the bare metal server to aaa cluster. I spoke with Arnaud and he will decide on a path for migration.

**Comment by BenTheElder** [comment]
We will need to raise this to
/priority important-soon
Given https://www.theregister.com/2024/11/18/equinix_ends_metal_iaas/

We have time, and other pressing work, but we need to be tracking moving it.

IMHO: while there are other options out there, we have a lot of links to this in issues etc, if it's cheap to operate I think we should keep it for now and just lift it over to AAA cluster or similar.

What's left for this one?

**Comment by BenTheElder** [comment]
https://cs-canary.k8s.io/ seems to have performance issues but also maybe isn't running the same version:

https://cs.k8s.io/?q=NodeStageVolume&i=nope&files=&excludeFiles=&repos= 
<img width="1032" alt="Screenshot 2024-12-03 at 3 24 56 PM" src="https://github.com/user-attachments/assets/fff114b8-c09a-42ba-b7cd-54819e0df781">


https://cs-canary.k8s.io/?q=NodeStageVolume&i=nope&files=&excludeFiles=&repos= 
<img width="1044" alt="Screenshot 2024-12-03 at 3 24 59 PM" src="https://github.com/user-attachments/assets/77fef1c4-d0e0-47e9-b71b-0d051991ca1a">

(note the response times)

**Comment by vielmetti** [comment]
Re-upping this as it came up in conversation today on the sig-infra call and because I am actively working on timelines and budgets for the Equinix Metal wind-down.

**Comment by BenTheElder** [comment]
We should check if performance is disk, CPU or memory, we have options (like switching AAA to larger nodes, using a faster disk type, etc) but that needs investigating.

**Comment by BenTheElder** [comment]
And then we should just rotate cs.k8s.io to point to the deployment at cs-canary.k8s.io and wind down the equinix machine.

**Comment by BenTheElder** [comment]
I tried the test in https://github.com/kubernetes/k8s.io/issues/2182#issuecomment-2515795607 again, this time cs-canary.k8s.io gave an error page eventually.

Looking at the pod logs (it is in aaa cluster in kubernetes-public GCP project) there are a lot of git fetch errors, and one replica is not ready:

<details>

```
2025-01-16 10:07:34.137 PST
Continuing...
2025-01-16 10:07:34.139 PST
2025/01/16 18:07:34 Failed to git reset [git reset --hard origin/main] at "/data/data/vcs-b90b1681de4da9db8d670ab238c7ed231fc1b8a2", see output below
2025-01-16 10:07:34.139 PST
fatal: ambiguous argument 'origin/main': unknown revision or path not in the working tree.
2025-01-16 10:07:34.139 PST
Use '--' to separate paths from revisions, like this:
2025-01-16 10:07:34.139 PST
'git <command> [<revision>...] -- [<file>...]'
2025-01-16 10:07:34.139 PST
: exit status 128
2025-01-16 10:07:34.139 PST
Continuing...
2025-01-16 10:07:34.994 PST
2025/01/16 18:07:34 Failed to git fetch [git fetch --prune --no-tags --depth 1 origin +main:remotes/origin/main] at "/data/data/vcs-191ecf511a99d78d8755448c91064a390b0762fe", see output below
2025-01-16 10:07:34.994 PST
fatal: couldn't find remote ref main
2025-01-16 10:07:34.994 PST
: exit status 128
2025-01-16 10:07:34.994 PST
Continuing...
2025-01-16 10:07:35.027 PST
2025/01/16 18:07:35 Failed to git reset [git reset --hard origin/main] at "/data/data/vcs-191ecf511a99d78d8755448c91064a390b0762fe", see output below
2025-01-16 10:07:35.028 PST
fatal: ambiguous argument 'origin/main': unknown revision or path not in the working tree.
2025-01-16 10:07:35.028 PST
Use '--' to separate paths from revisions, like this:
2025-01-16 10:07:35.028 PST
'git <command> [<revision>...] -- [<file>...]'
2025-01-16 10:07:35.028 PST
: exit status 128
2025-01-16 10:07:35.028 PST
Continuing...
2025-01-16 10:07:36.117 PST
2025/01/16 18:07:36 Failed to git fetch [git fetch --prune --no-tags --depth 1 origin +main:remotes/origin/main] at "/data/data/vcs-431070b759cc3f63b26e44d16ad8aeb7cd1fd4aa", see output below
2025-01-16 10:07:36.216 PST
fatal: couldn't find remote ref main
2025-01-16 10:07:36.216 PST
: exit status 128
2025-01-16 10:07:36.216 PST
Continuing...
2025-01-16 10:07:36.216 PST
2025/01/16 18:07:36 Failed to git reset [git reset --hard origin/main] at "/data/data/vcs-431070b759cc3f63b26e44d16ad8aeb7cd1fd4aa", see output below
2025-01-16 10:07:36.216 PST
fatal: ambiguous argument 'origin/main': unknown revision or path not in the working tree.
2025-01-16 10:07:36.216 PST
Use '--' to separate paths from revisions, like this:
2025-01-16 10:07:36.216 PST
'git <command> [<revision>...] -- [<file>...]'
2025-01-16 10:07:36.216 PST
: exit status 128
2025-01-16 10:07:36.216 PST
Continuing...
2025-01-16 10:07:37.228 PST
2025/01/16 18:07:37 Failed to git fetch [git fetch --prune --no-tags --depth 1 origin +main:remotes/origin/main] at "/data/data/vcs-2f5713f4bdffe87ea49b3f54b5a803caed48ba3c", see output below
2025-01-16 10:07:37.417 PST
fatal: couldn't find remote ref main
2025-01-16 10:07:37.417 PST
: exit status 128
2025-01-16 10:07:37.417 PST
Continuing...
2025-01-16 10:07:37.417 PST
2025/01/16 18:07:37 Failed to git reset [git reset --hard origin/main] at "/data/data/vcs-2f5713f4bdffe87ea49b3f54b5a803caed48ba3c", see output below
2025-01-16 10:07:37.417 PST
fatal: ambiguous argument 'origin/main': unknown revision or path not in the working tree.
2025-01-16 10:07:37.417 PST
Use '--' to separate paths from revisions, like this:
2025-01-16 10:07:37.417 PST
'git <command> [<revision>...] -- [<file>...]'
2025-01-16 10:07:37.417 PST
: exit status 128
2025-01-16 10:07:37.417 PST
Continuing...
2025-01-16 10:07:40.537 PST
2025/01/16 18:07:40 merge 0 files + mem
2025-01-16 10:07:41.150 PST
2025/01/16 18:07:41 43287889 data bytes, 7067142 index bytes
2025-01-16 10:07:41.150 PST
2025/01/16 18:07:41 Searcher started for etcd-io/raft
2025-01-16 10:07:42.145 PST
2025/01/16 18:07:42 merge 2 files + mem
2025-01-16 10:07:43.516 PST
2025/01/16 18:07:43 Failed to git fetch [git fetch --prune --no-tags --depth 1 origin +main:remotes/origin/main] at "/data/data/vcs-f34cbc91cc3c0e7599d0191326cbe17f95b59ade", see output below
2025-01-16 10:07:43.622 PST
fatal: couldn't find remote ref main
2025-01-16 10:07:43.622 PST
: exit status 128
2025-01-16 10:07:43.622 PST
Continuing...
2025-01-16 10:07:43.622 PST
2025/01/16 18:07:43 Failed to git reset [git reset --hard origin/main] at "/data/data/vcs-f34cbc91cc3c0e7599d0191326cbe17f95b59ade", see output below
2025-01-16 10:07:43.622 PST
fatal: ambiguous argument 'origin/main': unknown revision or path not in the working tree.
2025-01-16 10:07:43.622 PST
Use '--' to separate paths from revisions, like this:
```

</details>

I probably won't be able to dig further for a bit, but clearly the canary deployment needs some work before we can switch.

**Comment by dims** [comment]
Specs from current cs.k8s.io

16 CPU(s)
```
root@k8s-code-2022:~# lscpu
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         39 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  16
  On-line CPU(s) list:   0-15
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Xeon(R) E-2278G CPU @ 3.40GHz
    CPU family:          6
    Model:               158
    Thread(s) per core:  2
    Core(s) per socket:  8
    Socket(s):           1
    Stepping:            13
    CPU max MHz:         5000.0000
    CPU min MHz:         800.0000
    BogoMIPS:            6799.81
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm
                         constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3
                          sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault ep
                         b invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid
                          mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_
                         l1d arch_capabilities
Virtualization features:
  Virtualization:        VT-x
Caches (sum of all):
  L1d:                   256 KiB (8 instances)
  L1i:                   256 KiB (8 instances)
  L2:                    2 MiB (8 instances)
  L3:                    16 MiB (1 instance)
NUMA:
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-15
Vulnerabilities:
  Itlb multihit:         KVM: Mitigation: VMX disabled
  L1tf:                  Not affected
  Mds:                   Not affected
  Meltdown:              Not affected
  Mmio stale data:       Mitigation; Clear CPU buffers; SMT vulnerable
  Retbleed:              Mitigation; Enhanced IBRS
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl and seccomp
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:            Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
  Srbds:                 Mitigation; Microcode
  Tsx async abort:       Mitigation; TSX disabled
```

32 GB Memory:
```
root@k8s-code-2022:~# free -mh
               total        used        free      shared  buff/cache   available
Mem:            31Gi       7.0Gi       658Mi        11Mi        23Gi        29Gi
Swap:          1.9Gi        40Mi       1.9Gi
```

**Comment by dims** [comment]
opened a PR to bump cs-canary to 4 CPU + 16 GB Memory for now - https://github.com/kubernetes/k8s.io/pull/7695
